File: app/src/main/java/closer/vlllage/com/closer/pool/Pool.java
package closer.vlllage.com.closer.pool;

import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.Map;

public class Pool {

    private final Map<Class, PoolMember> members = new HashMap<>();

    Pool() {}

    public static TempPool tempPool() {
        return new TempPool();
    }

    @SuppressWarnings("unchecked")
    public <T extends PoolMember> T $(Class<T> member) {
        if (!members.containsKey(member)) {
            try {
                members.put(member, member.getConstructor().newInstance().setPool(this));
            } catch (IllegalAccessException | InstantiationException | InvocationTargetException | NoSuchMethodException e) {
                e.printStackTrace();
            }
        }

        return (T) members.get(member);
    }

    public <T extends PoolMember> T $set(T member) {
        if (members.containsKey(member.getClass())) {
            throw new IllegalStateException("Cannot $set member that already exists");
        }

        members.put(member.getClass(), member);

        return (T) members.get(member);
    }

    protected void end() {
        for (PoolMember member : members.values()) {
            member.onPoolEnd();
        }
    }
}

File: app/src/main/java/closer/vlllage/com/closer/pool/Pool.kt
package closer.vlllage.com.closer.pool

import java.lang.reflect.InvocationTargetException
import java.util.*

open class Pool internal constructor() {

    private val members = HashMap<Class<*>, PoolMember>()

    fun <T : PoolMember> `$`(member: Class<T>): T {
        if (!members.containsKey(member)) {
            try {
                members[member] = member.getConstructor().newInstance().setPool(this)
            } catch (e: IllegalAccessException) {
                e.printStackTrace()
            } catch (e: InstantiationException) {
                e.printStackTrace()
            } catch (e: InvocationTargetException) {
                e.printStackTrace()
            } catch (e: NoSuchMethodException) {
                e.printStackTrace()
            }

        }

        return members[member] as T
    }

    fun <T : PoolMember> `$set`(member: T): T {
        if (members.containsKey(member.javaClass)) {
            throw IllegalStateException("Cannot \$set member that already exists")
        }

        members[member.javaClass] = member

        return member
    }

    open fun end() {
        for (member in members.values) {
            member.onPoolEnd()
        }
    }

    companion object {

        fun tempPool(): TempPool {
            return TempPool()
        }
    }
}
